KAFKA-14799: Ignore source task requests to abort empty transactions

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-14799)

Since invoking `KafkaProducer::abortTransaction` when no transaction is currently open (via `KafkaProducer::beginTransaction`) causes an `IllegalStateException`, requests by source tasks to abort the current transaction when they have not returned any records from `SourceTask::poll` will currently cause the task to fail.

This patch gracefully handles that scenario by skipping the call to `KafkaProducer::abortTransaction` but emitting a `WARN`-level log message.

A basic unit test is added to ensure that we do not attempt to abort empty transactions.

A small bug in the internal API we use in `ExactlyOnceWorkerSourceTaskTest` for manipulating and awaiting source task poll contents is also fixed (two lines to change, two hours to figure out 😄).

As a follow-up item, we may want to consider using the `MockProducer` class in these and other unit tests since it may involve less work than manually mocking the `Producer` interface and should more-closely resemble the behavior of the `KafkaProducer` class (for example, it will automatically check for transaction state during calls to `beginTransaction`, `abortTransaction`, and `commitTransaction`).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-14799: Ignore source task requests to abort empty transactions

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java:
```diff
@@ -483,15 +483,19 @@ protected boolean shouldCommitTransactionForRecord(SourceRecord record) {
                             log.info("Aborting transaction for record on topic {} as requested by connector", record.topic());
                             log.trace("Last record in aborted transaction: {}", record);
-                            abortTransaction();
+                            maybeAbortTransaction();
```
just curios. Is it possible that the transaction is closed and abortable records are existent?

C0urante replied:
I don't think so; the only time we invoke `Producer::beginTransaction` is [here](https://github.com/apache/kafka/blob/ccfc389a638a126c3769bdd72725bae532ca4d01/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L248-L249), after which we immediately set `transactionOpen` to `true`, and the only time we set `transactionOpen` to `false` is after either aborting a transaction (which takes place in this method), or committing the transaction (which takes place [here](https://github.com/apache/kafka/blob/ccfc389a638a126c3769bdd72725bae532ca4d01/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L311). In either case, if we fail the operation (aborting/committing the transaction), the task fails, and we [skip the end-of-life offset commit](https://github.com/apache/kafka/blob/ccfc389a638a126c3769bdd72725bae532ca4d01/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L219-L222).

chia7712 replied:
Thanks for explanation!

+1 again

chia7712 reviewed (approved):
LGTM

one small question: Is following usage valid? and is it worth adding test case for it?
1. poll: r0, r1, r2
2. set commit on r2
3. set abort
4. expect all records get committed. nothing is abort

We don't check consistency between batch and record, so I guess it is allowed.

C0urante committed:
Add test cases for mixed record/batch transaction commit/abort, fix flaky unit test

C0urante committed:
Update TransactionContext Javadocs to clarify precedence for record- and batch-based operations

C0urante commented:
Thanks @chia7712, good catch with that edge case! You're correct about the behavior in that scenario; the record-based boundary is given precedence over the batch-based one.

I've added test cases to cover both instances of this scenario (record-based commit and batch-based abort, and batch-based commit and record-based abort), and I've updated the Javadocs for `TransactionContext` to clarify the behavior in this scenario.

I've also fixed a small flaky test issue in the `ExactlyOnceWorkerSourceTaskTest` suite.

C0urante commented:
Test failures are unrelated; merging.
